### PR TITLE
snap, store: filter out invalid snap edited links from store info and persisted state

### PR DESCRIPTION
**Pull request by ernestl**
**Issue:**
Store send out invalid empty EditedLinks that in newer snapd (e.g. v2.58, [after this change](https://github.com/snapcore/snapd/blob/master/snap/validate.go#L1167)) is introduced into snap state unvalidated. When reverting to snapd before this change, it causes disconnected interfaces because setup of security profiles fails when validating snap links as part of active snap info validation.

**Detail:**
The store send out empty EditedLinks which does not comply to the snap links [validation criteria](https://github.com/snapcore/snapd/blob/master/snap/validate.go#L1167). (A store fixed was released 11 July 2023).

snapd 2.57 (and earlier) [validates info.Links() ](https://github.com/snapcore/snapd/blob/release/2.57/snap/validate.go#L399) as part of 
refresh when setting up security profiles: ifacemgr StartUp -> snapsWithSecurityProfiles -> CurrentInfo -> addSnaps -> AddSnap -> Validate -> ValidateLinks

snapd 2.58 (and later) that includes [this change](https://github.com/snapcore/snapd/pull/12286/files#diff-291f552d73b5fd8ca18103ba5b54a36155556275d87bc73f9160086fc5e347d1R399) that introduce EditedLinks to snap info and
(1) [populates EditedLinks](https://github.com/snapcore/snapd/pull/12286/files#diff-c858d7cf81960ac83fef0a7b1c0da86a5cc71202a2f333f46323f8bde75a2081R274) from snap store information and 
(2) [does not validate](https://github.com/snapcore/snapd/pull/12286/files#diff-291f552d73b5fd8ca18103ba5b54a36155556275d87bc73f9160086fc5e347d1L399-R399) these EditedLinks when setting up security profiles

As a result:
-  If snaps with interfaces (security profiles) are/have been installed with snapd 2.58 with invalid EditedLinks received from the store (before the fix), refreshing to snapd 2.57 will result in those snaps fail security profile generation and result in disconnected interfaces to all related snaps. If it is core snap failing, all snaps connecting to "system slots" are disconnected and cannot be manually connected.

- On UC with core16, where snapd is embedded in core, after above up/down of core there will not be an issue, but then refreshing back to core with snapd 2.58 will fail and result in interfaces disconnected and manual connect not working until after snapd restart.

**Proposal:**

1. This change does not focus on correcting invalid state data
2. Normalize EditedLinks as part of function `(s *Info) Links()` to convert email links to `mailto:` and filter out invalid entries as used for cmd `snap info` (clientInfo) and for callers of `(s *Info) Contact()` and `(s *Info) Website()` etc.
3. Propose more aggressive validation to  for normalization helper `addLink` to ensure all normalized links will pass validation. This big implication here is that contacts without @ (basic email sanity check) will not make the cut anymore, hence the need to change unit tests.
4. Propose always testing for both OrignalLinks and EditedLinks as part of snap info validation [This idea was rejected]
5. Added normalization step of EditedLinks where received from the store. Arguably not required now that store links are fixed?
6. Propose concept of full info validation on store information [This idea was abandoned for now, needs more consideration]

**Open Questions:**

1. Does it make sense to re-validate snap.Info during doSetupProfiles (e.g. originating from snap install, enable, quota, autoconnect)?

There are two instances of snap.Info validation external to the snap module:
    - snap/pack/pack.go
    - interfaces/repo.go        <-- looking at this one

Call path: (m *InterfaceManager) doSetupProfiles -> m.setupProfilesForSnap -> m.(r *Repository).AddSnap -> snap.Validate

-  This this case snap.Info is retrieved in doSetupProfiles with call to snap.ReadInfo and passed down the call path. Arguable  validation only needs to happen at the source (snap.ReadInfo - and it does) and after additional modifications to the structure ([see example](https://github.com/snapcore/snapd/pull/12983/files/ab50df30d4305828879094c63bc464207e75f5c2#diff-56f28445609dd85879b5c353c23437fc57aecebd5c2f8143d61962976c0b5c25R72)) . One challenge is that efficiently passing structures by reference leaves it open to accidental or intentional modification between source and point of consumption (go does not have const equivalent to enforce read-only access).

- An additional concern is that snap.Validate does not always validate all fields e.g. in the latest code EditedLinks is not validated.

2. Does it make sense to still validate the edited links coming from the store given the store is now fixed, and generally store data is not validated at the source? If we want to transition to stricter validation at source, should we plan work for this?

### CODE REVIEW

File: store/store.go [hunk 1]
```diff
@@ -964,6 +964,10 @@ func (s *Store) SnapInfo(ctx context.Context, snapSpec SnapSpec, user *auth.User
 	if err != nil {
 		return nil, err
 	}
+	// TODO: Consider enabling this when the store code is fixed and sends us valid snap information
+	//if snap.Validation(info); err != nil {
+	//	return nil, err
+	//}
```
Comment (pedronis) [inline store/store.go:970]: I'm not sure we want to pursue this

File: snap/validate.go [hunk 2]
```diff
@@ -399,6 +399,10 @@ func Validate(info *Info) error {
 	if err := ValidateLinks(info.OriginalLinks); err != nil {
 		return err
 	}
+	// TODO: Consider adding once cleaning of persisted invalid EditedLinks is available
+	//if err := ValidateLinks(info.EditedLinks); err != nil {
+	//        return err
+	//}
```
Comment (pedronis) [inline snap/validate.go:405]: adding this is a bit odd wrt the normalisation code we added

**Review by pedronis — COMMENTED** [review]
I'm unsure about the TODOs

**Review by pedronis — COMMENTED** [review]
couple more considerations

File: store/store.go (hunk 1 above)
Comment (ernestl) [reply store/store.go:970]: Removed

File: store/details_v2.go [hunk 3]
```diff
@@ -282,6 +282,8 @@ func infoFromStoreSnap(d *storeSnap) (*snap.Info, error) {
 	// needs to be set for old snapd
 	info.LegacyEditedContact = d.Contact
 	info.EditedLinks = d.Links
+	// TODO: Consider removing this when the store code is fixed and sends us valid EditedLinks
+	info.EditedLinks = info.NormalizedEditedLinks()
```
[the remark this answers is not in the document]
Comment (ernestl) [reply store/details_v2.go:286]: Good idea, thanks

File: snap/validate.go (hunk 2 above)
Comment (ernestl) [reply snap/validate.go:405]: Removed

File: snap/info.go [hunk 4]
```diff
@@ -501,44 +501,83 @@ func (s *Info) Description() string {
...
+	if strutil.ListContains(links[key], link) {
+		return
+	}
+	links[key] = append(links[key], link)
+}
+
+// NormalizedEditedLinks returns the normalized edited links that adhere to
+// ValidateLinks validation criteria. Attempts to convert URL scheme "" 
+// links to "mailto" and avoids duplicate link entries.
+func (s *Info) NormalizedEditedLinks() map[string][]string {
```
[the remark this answers is not in the document]
Comment (ernestl) [reply snap/info.go:555]: I used your suggestion to use info.Links to normalise, so not required anymore.

File: snap/info_test.go
[the remark this answers is not in the document]
Comment (ernestl) [reply snap/info_test.go:1]: Done

File: store/details_v2.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (pedronis) [reply store/details_v2.go:285]: we could also write the code like this, which doesn't have the issue:

```
if len(d.Links) != 0 {
    info.EditedLinks = d.Links
    info.EditedLinks = info.Links()
}
```

which would be safe and a bit more efficient

File: snap/info_test.go [hunk 5]
```diff
@@ -152,41 +152,71 @@ func (s *infoSuite) TestContactFromLinksNotEmail(c *C) {
...
+				"":        {"ocontact2@example.com"},
+				"?":       {"ocontact3@example.com"},
+				"abc":     {},
+			},
+		},
+	}
+
+	c.Check(snap.ValidateLinks(info.EditedLinks), NotNil)
+	c.Check(snap.ValidateLinks(info.NormalizedEditedLinks()), IsNil)
+	c.Check(info.NormalizedEditedLinks(), DeepEquals, map[string][]string{
```
Comment (pedronis) [inline snap/info_test.go:194]: this can also use info.Links()

**Review by pedronis — COMMENTED** [review]
answered discussion

File: store/details_v2.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ernestl) [reply store/details_v2.go]: Thanks, implemented this and added comment to make it more obvious why the check is required.

File: snap/info_test.go (hunk 5 above)
Comment (ernestl) [reply snap/info_test.go]: Changed

**Review by pedronis — APPROVED** [review]
thank you

**Review by mvo5 — APPROVED** [review]
Thank you

**Pull request merged**